mysqltest/mysqltest.go

```diff
@@ -0,0 +1,44 @@
+// Copyright 2022 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mysqltest provides an ephemeral MySQL server for testing database integration. It's
+// designed to be used in code that needs to work inside and outside Google.
+package mysqltest
+
+import (
+	"io"
+	"testing"
+)
+
+// ConnInfo specifies how to connect to the MySQL server that is created by this package.
+type ConnInfo struct {
+	Username string
+	Password string
+	Hostname string
+	Port     string
+}
+
+// MustStart starts a MySQL server, or panics if there was an error.
+func MustStart(t *testing.M, opts ...Option) (ConnInfo, io.Closer) {
```

Comment:
I wonder if we can just call `t.Cleanup(func() {closer.Close()})` inside this func?

Reply:
I actually made a mistake here. This function should not accept *testing.T as a parameter, because it needs to run inside TestMain() functions which don't have a *testing.T. So I have to go in the opposite direction of what you're suggesting, regrettably.

Background: why does this function need to run in TestMain()? Because the common use case is to start one Docker instance and share it between all tests. The normal Go way to do this is to initialize the Docker instance in TestMain.

Reply:
Haha, fair enough.